### PR TITLE
Implement chemistry domain sub-package.

### PR DESCRIPTION
**Pull request by amccaskey**
# Overview

Implement support for generating molecular hamiltonians for common quantum chemistry use cases. 

Address #100 

Specifically: 

* `molecule.h`: common data types to describe molecular geometries, one and two body integrals, and molecular hamiltonians. Provide function in C++ to generate a molecular hamiltonian. 
* `MoleculePackageDriver.h`: Common interface for interfacing common chemistry packages that provide the ability to create molecular hamiltonians. 
* `uccsd.h`: CUDA Quantum kernel definitions for the unitary coupled cluster ansatz. 
* `PySCFDriver.h`: Implementation (as a plugin) of the `MoleculePackageDriver` interface which generates hamiltonians via PySCF using Pybind11's embedded interpreter. 
* `python/domains/chemistry.py`: python module providing a creation function for molecular hamiltonians which delegates to OpenFermion-PySCF.
* Update to `nvq++` to search for plugin libraries and link them properly.
* Tests for all the above. 

To make it easier to describe and extract information out of the one and two body integrals produced by the `MoleculePackageDriver` I have added `xtensor` as a new git submodule (BSD-3-Clause). This package provides Numpy-like data types in C++ for general N-rank tensor data.

### CODE REVIEW

File: python/domains/chemistry.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                                  charge=0,
+                                  n_active_electrons=None,
+                                  n_active_orbitals=None):
+    '''
+    Create the molecular hamiltonian corresponding to the provided 
+    geometry, basis set, multiplicity, and charge.  One can also specify the 
+    number of active electrons and orbitals, thereby approximating the 
+    molecular hamiltonian and freezing core orbitals. This function delegates 
+    to the OpenFermion-PySCF package and will throw an error if that module is 
+    not available
```
Comment (schweitzpgi) [inline python/domains/chemistry.py:34]:
```suggestion
    not available.
```

File: python/domains/chemistry.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    Args: 
+      geometry: The geometry should be provided as a list of tuples, 
+        where each tuple element contains the atom name and a tuple
+        of atom coordinates, e.g. [('H', (0.,0.,0.)), ('H', (0.,0.,.7474))].
+      basis: The basis set as a string
+      multiplicity: The spin multiplicity as an int
+      charge: The total charge of the molecular system as an int
```
Comment (schweitzpgi) [inline python/domains/chemistry.py:42]:
```suggestion
      basis: The basis set as a string.
      multiplicity: The spin multiplicity as an int.
      charge: The total charge of the molecular system as an int.
```

Comment (schweitzpgi) [reply python/domains/chemistry.py:42]: There are other missing periods in the sequent.

File: runtime/cudaq/domains/chemistry/MoleculePackageDriver.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  /// @brief Return a molecular_hamiltonian described by the given
+  /// geometry, basis set, multiplicity, and charge. Optionally
+  /// restrict the active space.
+  virtual molecular_hamiltonian
+  createMolecule(const molecular_geometry &geometry, const std::string &basis,
+                 int multiplicity, int charge,
+                 std::optional<std::size_t> nActiveElectrons = std::nullopt,
+                 std::optional<std::size_t> nActiveOrbitals = std::nullopt) = 0;
+};
+} // namespace cudaq
```
Comment (schweitzpgi) [inline runtime/cudaq/domains/chemistry/MoleculePackageDriver.h:30]: Looks like most files don't end in a newline?

File: runtime/cudaq/domains/chemistry/molecule.cpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+std::complex<double> &two_body_integals::operator()(std::size_t p,
+                                                    std::size_t q,
+                                                    std::size_t r,
+                                                    std::size_t s) {
+  return xt::adapt(data.get(), shape[0] * shape[1] * shape[2] * shape[3],
+                   xt::no_ownership(), shape)(p, q, r, s);
+}
+
+void two_body_integals::dump() {
+  std::cout << xt::adapt(data.get(), shape[0] * shape[1] * shape[2] * shape[3],
```
Comment (schweitzpgi) [inline runtime/cudaq/domains/chemistry/molecule.cpp:53]:
```suggestion
  std::cerr << xt::adapt(data.get(), shape[0] * shape[1] * shape[2] * shape[3],
```

File: tools/nvqpp/nvq++.in [hunk 5]
```diff
@@ -174,13 +174,27 @@ NVQPP_LD_PATH=${NVQPP_LD_PATH:-"@NVQPP_LD_PATH@"}
...
-LINKER_FLAGS="${LINKER_FLAGS} -Wl,-rpath,${install_dir}/lib -Wl,-rpath,${PWD}"
+LINKER_FLAGS="${LINKER_FLAGS} -Wl,-rpath,${install_dir}/lib -Wl,-rpath,${install_dir}/lib/plugins -Wl,-rpath,${PWD}"
 
 LIBRARY_MODE_EXECUTION_MANAGER="qir"
 PLATFORM_LIBRARY="default"
 LLVM_QUANTUM_TARGET="qir"
-LINKDIRS="-L${install_dir}/lib @CUDAQ_CXX_NVQPP_LINK_STR@"
-LINKLIBS="-lcudaq -lcudaq-common -lcudaq-mlir-runtime -lcudaq-builder -lcudaq-ensmallen -lcudaq-nlopt -lcudaq-spin"
+LINKDIRS="-L${install_dir}/lib -L${install_dir}/lib/plugins @CUDAQ_CXX_NVQPP_LINK_STR@"
+LINKLIBS="-lcudaq -lcudaq-common -lcudaq-mlir-runtime -lcudaq-builder -lcudaq-ensmallen -lcudaq-nlopt -lcudaq-spin -lcudaq-chemistry"
```
Comment (schweitzpgi) [inline tools/nvqpp/nvq++.in:183]: Should this always be included? It sounds like it is pretty narrowly focused with the name "chemistry"?

File: python/CMakeLists.txt [hunk 6]
```diff
@@ -54,6 +55,7 @@ target_link_libraries(_pycudaq PUBLIC nvqir
 
 install(TARGETS _pycudaq DESTINATION ${CMAKE_INSTALL_PREFIX})
 install(FILES cudaq.py DESTINATION ${CMAKE_INSTALL_PREFIX})
+install(FILES apps/chemistry.py DESTINATION ${CMAKE_INSTALL_PREFIX}/apps/)
```
Comment (bettinaheim) [inline python/CMakeLists.txt:58]: This causes the build to fail due to an incorrect path during installation (CUDA Quantum can be built and tested successfully, but won't be installed correctly, hence the unit tests passing but the packaging validation failing):

```suggestion
install(FILES domains/chemistry.py DESTINATION ${CMAKE_INSTALL_PREFIX}/apps/)
```

File: tools/nvqpp/nvq++.in (hunk 5 above)
Comment (schweitzpgi) [reply tools/nvqpp/nvq++.in:183]: That sounds good. Add an option that pulls in the extra headers and libraries semi-magically. I like it.

Comment (amccaskey) [reply tools/nvqpp/nvq++.in]: Done.

**Review by schweitzpgi — APPROVED** [review]

**Pull request merged**